### PR TITLE
perf(projects): improve get_access_by_project speed by not serializing teams

### PR DESCRIPTION
**Pull request by JoshFerge**
according to this profile, we spend a lot of time serializing in `get_access_by_project`. ~600ms in large project queries

https://sentry.sentry.io/profiling/profile/sentry/e409e184a7a7449c883c90abcdcc28f3/flamegraph/?colorCoding=by+system+vs+application+frame&fov=10670915%2C96%2C3354361600%2C10&query=&sorting=call+order&tid=136213193283328&type=flamechart&view=top+down

This PR aims to remove this serialization by instead only selecting the project_id/team_ids in the django query, and only using ids in subsequent queries.

Since the models aren't actually used, we don't need all of the models data, and simply their ids.

### CODE REVIEW

File: src/sentry/api/serializers/models/project.py [hunk 1]
```diff
@@ -98,13 +97,15 @@ def get_access_by_project(
...
 
-    project_teams = list(ProjectTeam.objects.filter(project__in=projects).select_related("team"))
-    project_team_map = defaultdict(list)
+    project_teams = ProjectTeam.objects.filter(project__in=projects).values_list(
+        "project_id", "team_id"
+    )
 
-    for pt in project_teams:
-        project_team_map[pt.project_id].append(pt.team)
+    project_to_teams = defaultdict(list)
```
Comment (anonrig) [inline src/sentry/api/serializers/models/project.py:104]: Can you explicitly define the type of this variable?

File: src/sentry/api/serializers/models/project.py [hunk 2]
```diff
@@ -98,13 +97,15 @@ def get_access_by_project(
...
+    for project_id, team_id in project_teams:
+        project_to_teams[project_id].append(team_id)
 
-    team_memberships = _get_team_memberships([pt.team for pt in project_teams], user)
+    team_memberships = _get_team_memberships([team_id for _, team_id in project_teams], user)
```
Comment (anonrig) [inline src/sentry/api/serializers/models/project.py:108]: you can construct this array in 3 lines before where we already iterate it in a for loop. therefore, we wouldn't iterate the items twice

File: src/sentry/api/serializers/models/project.py [hunk 3]
```diff
@@ -98,13 +97,15 @@ def get_access_by_project(
 ) -> MutableMapping[Project, MutableMapping[str, Any]]:
     request = env.request
 
-    project_teams = list(ProjectTeam.objects.filter(project__in=projects).select_related("team"))
-    project_team_map = defaultdict(list)
+    project_teams = ProjectTeam.objects.filter(project__in=projects).values_list(
```
Comment (anonrig) [inline src/sentry/api/serializers/models/project.py:100]: This function takes sequence of projects. Is it possible to reduce this into sequence of project ids as well?

Comment (JoshFerge) [reply src/sentry/api/serializers/models/project.py:100]: there's no performance gain in doing so as we're not deserializing anything extra at the moment here

File: src/sentry/api/serializers/models/project.py (hunk 1 above)
Comment (JoshFerge) [reply src/sentry/api/serializers/models/project.py:104]: we do need to get typing enabled for this file, i can do so in a follow up.

**Review by anonrig — APPROVED** [review]

File: src/sentry/api/serializers/models/project.py [hunk 4]
```diff
@@ -81,7 +80,7 @@
 }
 
 
-def _get_team_memberships(team_list: Sequence[Team], user: User) -> Iterable[int]:
+def _get_team_memberships(team_list: Sequence[int], user: User) -> Iterable[OrganizationMemberTeam]:
```
Comment (markstory) [inline src/sentry/api/serializers/models/project.py:83]: Is there a benefit to changing the lookup used here to be `team_id__in` instead of `team_in`?

**Review by markstory — APPROVED** [review]
Looks good to me. Avoiding a join and only loading the necessary attributes will help improve query result generation time.

**Pull request merged**